FavoriteGameList: Add initial version

gh215 opened the pull request.

gh215 committed:
FavoriteGameList: Add initial version

amavrin commented on chapter-4/FavotireGameList/FavotireGameList/FavotireGameList.cpp:
```diff
@@ -0,0 +1,125 @@
+	cout << "Выберите " << WATCH << ", чтобы посмотреть список текущих игр" << endl;
+	cout << "Выберите " << ADD << ", чтобы добавить игру" << endl;
+	cout << "Выберите " << DELETE << ", чтобы удалить игру" << endl;
+	cout << "Выберите " << DELETE_ALL << ", чтобы удалить все игры" << endl;
```
При выборе 5, получаем сообщение:
```
Ваш выбор: 5

Неверный ввод, попробуйте ещё раз
```

amavrin commented on chapter-4/FavotireGameList/FavotireGameList/FavotireGameList.cpp:
```diff
@@ -0,0 +1,125 @@
+	cout << "\nЧто бы вы хотели сделать?" << endl;
+	cout << "Выберите " << WATCH << ", чтобы посмотреть список текущих игр" << endl;
+	cout << "Выберите " << ADD << ", чтобы добавить игру" << endl;
+	cout << "Выберите " << DELETE << ", чтобы удалить игру" << endl;
```
Если попросить удалить игру, которой нет в списке, не будет сообщения об ошибке

amavrin commented on chapter-4/FavotireGameList/FavotireGameList/FavotireGameList.cpp:
```diff
@@ -0,0 +1,125 @@
+
+	enum actions { WATCH = 1, ADD, DELETE, DELETE_ALL, QUIT };
+
+	cout << "Добро пожаловать в нашу программу по спику ваших любимых игр" << endl;
```
нет опции получить help по требованию

amavrin commented on chapter-4/FavotireGameList/FavotireGameList/FavotireGameList.cpp:
```diff
@@ -0,0 +1,125 @@
+	cout << "Добро пожаловать в нашу программу по спику ваших любимых игр" << endl;
+
+	cout << "\nЧто бы вы хотели сделать?" << endl;
+	cout << "Выберите " << WATCH << ", чтобы посмотреть список текущих игр" << endl;
```
не хватает переноса строки перед первой игрой:
```
На данный момент у вас 2 игр и среди них: morskoy-boy
krestiki-noliki
```

amavrin commented on chapter-4/FavotireGameList/FavotireGameList/FavotireGameList.cpp:
```diff
@@ -0,0 +1,125 @@
+	enum actions { WATCH = 1, ADD, DELETE, DELETE_ALL, QUIT };
+
+	cout << "Добро пожаловать в нашу программу по спику ваших любимых игр" << endl;
+
```
Если ввести "q", программа зациклится:
```
то бы вы хотели сделать?
Выберите 1, чтобы посмотреть список текущих игр
Выберите 2, чтобы добавить игру
Выберите 3, чтобы удалить игру
Выберите 4, чтобы удалить все игры
Если хотите выйти из программы - введите 5

Ваш выбор: q
Неверный ввод, попробуйте ещё раз

Ваш выбор:
Неверный ввод, попробуйте ещё раз

Ваш выбор:
Неверный ввод, попробуйте ещё раз

Ваш выбор:
Неверный ввод, попробуйте ещё раз
```

amavrin commented on chapter-4/FavotireGameList/FavotireGameList/FavotireGameList.cpp:
```diff
@@ -0,0 +1,125 @@
+				{
+					cout << *watchGames << endl;
+				}
+			}
```
Можно уменьшить вложенность, если использовать оператор `continue`:
```suggestion
			if (gameList.empty())
			{
				cout << "\nИзвините, но в вашей библиотеке нет игр" << endl;
				continue;
			}
			cout << "\nНа данный момент у вас " << gameList.size() << " игр и среди них: ";
			for (watchGames = gameList.begin(); watchGames != gameList.end(); ++watchGames)
			{
				cout << *watchGames << endl;
			}

```

amavrin reviewed (commented):
В целом очень неплохо, нужно исправить некоторые шероховатости и научиться вводить игры с пробелами.

gh215 committed:
FavoriteGameList: Removed bug with infinite loop, added new HELP function,
described QUIT function and added line break for games.

Pseudocode: moved to txt from docx

amavrin commented on chapter-4/FavotireGameList/FavotireGameList/FavotireGameList.cpp:
```diff
@@ -0,0 +1,156 @@
+	int user_input;
+	string user_game_input;
+
+	enum actions { WATCH = 1, ADD, DELETE, DELETE_ALL, HELP, QUIT };
```
QUIT лучше ставить первым, чтобы выходу соответствовал '0'

amavrin commented on chapter-4/FavotireGameList/FavotireGameList/FavotireGameList.cpp:
```diff
@@ -0,0 +1,156 @@
+	cout << "Выберите " << DELETE << ", чтобы удалить игру" << endl;
+	cout << "Выберите " << DELETE_ALL << ", чтобы удалить все игры" << endl;
+	cout << "Выберите " << HELP << ", чтобы получить помощь" << endl;
+	cout << "Если хотите выйти из программы - введите " << QUIT << endl;
```
Можно ввести 7, и программа тихо это съест - ни ошибки, ни действия

amavrin commented on chapter-4/FavotireGameList/FavotireGameList/FavotireGameList.cpp:
```diff
@@ -0,0 +1,156 @@
+					cout << " " << *watchGames << endl;
+				}
+			}
+			break;
```
Мы говорили, что такую конструкцию можно делать более читаемой, уменьшив вложенность:

```suggestion
			if (gameList.empty())
			{
				cout << "\nИзвините, но в вашей библиотеке нет игр" << endl;
				break;
			}
			
			cout << "\nНа данный момент у вас " << gameList.size() << " игр и среди них: \n";
			for (watchGames = gameList.begin(); watchGames != gameList.end(); ++watchGames)
			{
				cout << " " << *watchGames << endl;
			}
			break;
```

amavrin reviewed (commented):
Стало лучше!
Но нужно
1. не потерять проверку на неверно введённый символ - например, 7
2. оптимизировать if-else

Также, очень рекомендую использовать функции. Напрашивается несколько:
* вывести справку по игре `void print_help()`
* вывести список игр `void print_game_list(vector<string>gameList);`
* поиск конкретной игры `bool is_game_present(vector<string>gameList, string game)`
* добавление игры `void add_game(vector<string>& gameList, string game)`
* удаление игры `bool delete_game(vector<string>& gameList, string game)`

Использование функций очень важно.

И интересно, а что не получилось в том, чтобы вводить игру с пробелами?

gh215 committed:
Temp.

gh215 committed:
FavoriteGameList: Updated pseudocode of the program, added condition if user entered wrong digit value

amavrin reviewed (commented):
Лишние `else` в коде не убраны, и нет unit-тестов

gh215 committed:
FavoriteGameList: temp.

gh215 committed:
FavoriteGameList: temp.

gh215 committed:
FavoriteGameList: temp.

gh215 committed:
FavoriteGameList: temp.

gh215 committed:
FavoriteGameList: Added test-cases, added new functions

gh215 committed:
FavoriteGameList: refactor real_add_game and real_delete_game

gh215 committed:
FavoriteGameList: Final version of program
